fc/parsing/actions.py

```diff
@@ -941,60 +941,77 @@ def _expr(self):
 
 
 class UnitRef(BaseGroupAction):
-    """Parse action for unit references within units definitions."""
+    """Parse action for unit references within units definitions.
 
-    def get_value(self, token, negate=False):
-        """Get a decent string representation of the value of the given numeric token.
-        It may be a plain number, or it may be a simple expression which we have to evaluate.
-        """
-        format = "%.17g"
-        result = str(token).strip()
-        try:
-            value = float(result)
-        except ValueError:
-            # Evaluation required; somewhat risky!
-            value = eval(result)
-            if negate:
-                value = -value
-            result = format % value
-        else:
-            # Just use the string representation in the protocol
-            if negate:
-                if result[0] == '-':
-                    result = result[1:]
-                else:
-                    result = '-' + result
-        return result
+    Properties:
 
-    # Leaving old XML method in to document existing properties / tokens.
-    # def _xml(self):
-    #    attrs = self.get_attribute_dict('prefix', 'units', 'exponent')
-    #    if 'multiplier' in self.tokens:
-    #        attrs['multiplier'] = self.GetValue(self.tokens['multiplier'][0])
-    #    if 'offset' in self.tokens:
-    #        attrs['offset'] = self.GetValue(self.tokens['offset'][0][1], self.tokens['offset'][0][0] == '-')
-    #    return CELLML.unit(**attrs)
+    ``units``
+        Name of the units referenced.
+    ``prefix``
+        Optional SI prefix name, e.g. milli, deca. Defaults to ''.
+    ``multiplier``
+        Optional scalar multiplier; None if omitted.
+    ``exponent``
+        Optional exponent; None if omitted.
+    ``offset``
+        For defining units such as fahrenheit. Will raise an error if given.
+    ``pint_expression``
+        Pint syntax for this unit reference.
+    """
+    def _expr(self):
+        if 'offset' in self.tokens:
+            raise ValueError('Offset units are no longer supported')
+        self.units = self.get_named_token_as_string('units')
+        assert self.units is not None
+        self.prefix = self.get_named_token_as_string('prefix', '')
+        self.exponent = self.get_named_token_as_string('exponent')
+        self.multiplier = self.get_named_token_as_string('multiplier')
+
+        expr = self.prefix + self.units
+        if self.exponent is not None:
+            expr = '{}**{}'.format(expr, self.exponent)
```

Comment:
Do any of these need brackets?

Reply:
I don't /think/ so, precedence should make it work...